iteratee/src/main/scala/io/finch/iteratee/package.scala

```diff
@@ -68,19 +68,18 @@ trait IterateeInstances {
     private def writeIteratee(w: Writer[Buf]): Iteratee[F, Buf, Unit] =
       Iteratee.foreachM[F, Buf](chunk => TE(w.write(chunk)))
 
-    def apply(s: Enumerator[F, A], cs: Charset): Reader[Buf] = {
+    def apply(s: Enumerator[F, A], cs: Charset): F[Reader[Buf]] = {
       val p = new Pipe[Buf]
       val run = s
         .ensure(F.suspend(TE(p.close())))
         .map(chunk => encodeChunk(chunk, cs))
         .into(writeIteratee(p))
 
-      F.toIO(run).unsafeRunAsyncAndForget()
-      p
+      F.guarantee(F.pure(p: Reader[Buf]))(F.delay(F.toIO(run).unsafeRunAsyncAndForget()))
```

Comment:
This is the relevant piece. As we can't just `flatMap` from `run` (it only resolves stream terminates), we need a way to asynchronously mix an effect.

I'm wondering if there is a better way to do that (looking for suggestions).

Reply:
You could `F.runAsync(run)(_ => IO.unit).to[F]`. But I'm not sure I understand the goal here. Is it supposed to guarantee that everything gets encoded into `p` even on cancellation? `F.pure` can't fail so cancellation is the only thing left.

Reply:
Oh yeah, very good point about cancelation. I guess I'd need to use `guaranteeCase` to handle that?

Reply:
To answer your question: we previously unconditionally evaluated a stream when we call `ToResponse` (from within `ToService`) We want to change that and stay within `F[_]` as long as possible, and run everything together.

The cancelation part is really interesting. I'd imagine to properly support it here, we need to require `ConcurrentEffect` so we can `runCancellable`? Otherwise, I'm not sure the cancelation signal will get propagated via `runAsync`.